File: src/za/co/adhd_developers/java/questions/Question30.java
package za.co.adhd_developers.java.questions;

import za.co.adhd_developers.java.tools.Question;

public class Question30 implements Question {

    private int start = 2;
    private int end = 10000000;
    private int pow = 5;

    private int answer = 0;

    @Override
    public void doWork() {
        for (int i = this.start; i < this.end; i++) {
            String number = String.valueOf(i);

            String digits[] = number.split("|");

            int sum = 0;

            for (int j = 0; j < digits.length; j++) {
                sum += Math.pow(Integer.valueOf(digits[j]), this.pow);
            }

            if (sum == i) {
                this.answer += sum;
            }
        }
    }

    @Override
    public void printAnswer() {
        System.out.println("=================================");
        System.out.println("Question 30");
        System.out.println("The sum of all numbers that can be written as the sum of fifth powers of their digits, is: " + this.answer);
        System.out.println("=================================");
    }
}

File: src/za/co/adhd_developers/kotlin/questions/Question30.kt
package za.co.adhd_developers.kotlin.questions

import za.co.adhd_developers.kotlin.Question

class Question30 : Question {

    val powAnswers = HashMap<String, Double>()

    val powTo = "5"

    var answer = 0.0

    override fun doWork() {
        for (i in 0 until 10) {
            val pow = Math.pow(i.toDouble(), powTo.toDouble())

            powAnswers[i.toString()] = pow
        }

        val min = 2
        val max = 10000000

        for (n in min until max) {
            var digits = n.toString().toCharArray().map { c -> c.toString() }

            var possible = 0.0

            digits.forEach { s: String -> possible = possible.plus(powAnswers[s]!!) }

            if (possible == n.toDouble()) {
                answer = answer.plus(possible)
                println(answer)
            }
        }
    }

    override fun printAnswer() {
        println("=================================")
        println("Question 30")
        println("The sum of all numbers that can be written as the sum of fifth powers of their digits, is: " + this.answer)
        println("=================================")
    }
}